### ISSUE TITLE
SC2218 when overriding existing function

### ISSUE THREAD

**Issue by Toreno96** [koalaman/shellcheck#1595]
#### For bugs
- Rule Id (if any, e.g. SC1000): [SC2218](https://github.com/koalaman/shellcheck/wiki/SC2218)
- My shellcheck version (`shellcheck --version` or "online"): 0.6.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

File `b.sh`:
```bash
#!/bin/bash

foo() {
    echo 'foo'
}
foo

foo() {
    echo 'overridden'
}
foo
```

#### Here's what shellcheck currently says:

```
$ shellcheck b.sh

In b.sh line 6:
foo
^-^ SC2218: This function is only defined later. Move the definition up.

For more information:
  https://www.shellcheck.net/wiki/SC2218 -- This function is only defined lat...
```

#### Here's what I wanted or expected to see:

No error, because it is not true, that "this function is _only_ defined later", `b.sh` works as expected:
```
$ ./b.sh
foo
overridden
```

Optionally, there could be a warning that I'm overriding existing function.

**Comment by wileyhy** [comment]
Encountering this SC issue while trying to isolate a double free in bash.

The way Shellcheck's error is worded isn't factually accurate. SC says, "only defined later," which in this case isn't true. In my opinion, it's something concerning this idea of, "can I trust ShellCheck to give me accurate information?"  Perhaps it would be more accurate for ShellCheck to say something like, "This function is defined later. Please consider moving the definition up." 

> In ./test-script.sh_2 line 28:
> test; echo "$?"
> ^--^ SC2218 (error): This function is only defined later. Move the definition up.
> 

[liveuser@localhost-live]$ `vim ./test-script.sh_2` 

> ...
>  13 test(){
>  14   shopt -s expand_aliases;
>  15   echo "$ec", shopt: "$?"
>  16   alias M_='m=("${l[@]}")';
>  17   echo "$ec", alias: "$?"
>  18   wait -f;
>  19   echo "$ec", wait: "$?"
>  20   type -a M_;
>  21   echo "$ec", type: "$?"
>  22   M_;
>  23   echo "$ec", alias:M_: "$?"
>  24   };
>  25 echo "$ec", definition of function:test "$?"
>  26 type -a test;
>  27 echo "$ec", type "$?"
>  28 test;
>  29 echo "$ec", function:test "$?"
>  30 M_
>  31 echo "$ec", alias:M_ "$?"
>  32 set -x;
>  33 test(){
>  34   shopt -s expand_aliases;
>  35   echo "$ec", shopt: "$?";
>  36   alias M_='m=("${l[@]}")';
>  37   echo "$ec", alias: "$?";
>  38   wait -f;
>  39   echo "$ec", wait: "$?";
>  40   type -a M_;
>  41   echo "$ec", type: "$?";
>  42   M_;
>  43   echo "$ec", alias:M_: "$?";
>  44   };
>  45 echo "$ec", duplicate re-definition of function:test "$?"
> ...

Thank you
